# Image Classification dengan Vision Transformer (ViT)

Notebook ini mendemonstrasikan cara menggunakan model AI untuk klasifikasi gambar. Kita akan memuat gambar dari URL, menjalankan klasifikasi menggunakan pipeline dari Hugging Face Transformers, dan menampilkan hasilnya.

## Import Required Libraries
Import necessary libraries including transformers and IPython for image display.

In [ ]:
%pip install Pillow transformers torch

In [1]:
from transformers import pipeline
from IPython.display import Image

/Users/zeihanaulia/Programming/research/agent/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


# Load and Display Image
Load the image from the URL and display it in the notebook.

In [7]:
image_url = "https://huggingface.co/datasets/Narsil/image_dummy/raw/main/parrots.png"
Image(url=image_url)

# Create Image Classification Pipeline
Initialize the image classification pipeline using the specified model from transformers.

In [8]:
classifier = pipeline("image-classification", model="google/vit-base-patch16-224")

Device set to use mps:0


# Penjelasan Cara Kerja Pipeline dan Model
Baris kode `classifier = pipeline("image-classification", model="google/vit-base-patch16-224")` membuat sebuah "pipeline" siap pakai untuk klasifikasi gambar menggunakan model Vision Transformer (ViT) dari Google.

### Apa itu Pipeline?
- **Pipeline** adalah alat dari library `transformers` yang menyederhanakan penggunaan model AI. Ia menggabungkan beberapa langkah: memuat model, preprocessor (untuk gambar), dan postprocessor (untuk hasil).
- Dengan satu baris kode, Anda bisa langsung menggunakan model tanpa perlu setup manual.

### Apa itu Model ViT (Vision Transformer)?
- **ViT** adalah model AI modern yang menggunakan arsitektur "Transformer" (awalnya untuk teks, seperti GPT), tapi diadaptasi untuk gambar.
- **Patch-based processing**: Gambar dibagi menjadi "patch" kecil (16x16 pixel), lalu diubah menjadi token seperti kata dalam teks. Model belajar pola dari patch-patch ini.
- **Tidak seperti CNN lama**: ViT tidak menggunakan convolutional layers, tapi self-attention untuk memahami hubungan antar patch.

### Bagaimana Model Bisa "Menebak" (Prediksi)?
1. **Pelatihan (Training)**: Model dilatih pada dataset besar seperti ImageNet (1 juta+ gambar, 1000+ kelas). Ia belajar mengenali pola: misalnya, "paruh melengkung + bulu berwarna" = burung.
2. **Inference (Prediksi)**: Saat diberi gambar baru:
   - Gambar diproses menjadi patch dan token.
   - Model menghitung "attention" (fokus pada bagian penting, seperti kepala burung).
   - Output probabilitas untuk setiap kelas (misalnya, 99% macaw, 0.5% toucan).
   - Prediksi berdasarkan pola yang dipelajari, bukan "menebak" acak—tapi bisa salah jika gambar tidak mirip training data.

### Mengapa Akurat?
- Model ViT ini sudah fine-tuned (disesuaikan) untuk klasifikasi gambar, dengan akurasi tinggi (~80-90% pada ImageNet).
- Untuk gambar parrots, ia menebak "macaw" karena pola cocok dengan data training.

# Perform Image Classification
Run the classifier on the loaded image to obtain classification results.

In [4]:
result = classifier(image_url)

# Display Classification Results
Print or visualize the classification results in a readable format within the notebook.

In [5]:
for item in result:
    print(f"Label: {item['label']}, Score: {item['score']:.4f}")

Label: macaw, Score: 0.9905
Label: African grey, African gray, Psittacus erithacus, Score: 0.0056
Label: toucan, Score: 0.0011
Label: sulphur-crested cockatoo, Kakatoe galerita, Cacatua galerita, Score: 0.0007
Label: lorikeet, Score: 0.0007


# Penjelasan Hasil Klasifikasi
Hasil di atas menunjukkan prediksi model terhadap gambar yang dimuat. Berikut cara membacanya:

- **Label**: Nama objek atau kelas yang diprediksi (dalam bahasa Inggris, berdasarkan dataset ImageNet).
- **Score**: Probabilitas atau confidence level (0 hingga 1). Semakin tinggi, semakin yakin model bahwa gambar cocok dengan label tersebut.

### Interpretasi Hasil:
- **Prediksi utama**: `macaw` dengan score 0.9905 (99.05% confidence). Artinya, model sangat yakin gambar adalah burung macaw (jenis burung beo).
- Prediksi lainnya: Model juga mempertimbangkan kemungkinan kecil untuk jenis burung lain seperti African grey parrot, toucan, dll., tapi dengan score jauh lebih rendah (<1%).

Hasil ini akurat untuk gambar burung parrots, karena macaw adalah jenis burung yang mirip. Jika gambar berbeda, hasil akan bervariasi. Anda bisa ganti URL gambar di cell sebelumnya untuk eksperimen!